posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,196 @@
+    </i>
+</p>
+
+## Array API dispatching system
+
+Array API is a standard specification of functionality that array libraries need to implement. There is a [dedicated test suite](https://github.com/data-apis/array-api-tests) to verify the compliance of specific implementations with the standard, so it's easy to check all libraries adopting the standard to have a unified behavior. NEP 47 has a couple of examples on [usage of the Array API namespace](https://numpy.org/neps/nep-0047-array-api-standard.html#adoption-in-downstream-libraries).
+
+`__array_namespace__` is a central piece of the Array API dispatch method. It is a method of an array object that returns an object that has all the array API functions on it. In the case of NumPy calling `__array_namespace__() ` would return `numpy.array_api` module. The `__array_namespace__` method attached to the array object enables downstream libraries to consume multiple kinds of arrays, without having to have a hard dependency on any of those array libraries.
+The pattern to support multiple array libraries is intended to be similar to the following code snippet:
+```python
+def somefunc(x, y):
+    # Retrieves standard namespace.
+    # Raises if x and y have different namespaces.
+    xp = get_namespace(x, y)
+    out = xp.mean(x, axis=0) + 2*xp.std(y, axis=0)
+    return out
+```
+
+Notice that to make the above code valid, we don't need to import the package that implements mean and std functions explicitly; the required module is stored in x and y variables.
+
+For several concrete use cases of the Array API check out [the dedicated page](https://data-apis.org/array-api/latest/use_cases.html) of the standard. Anirudh Dagar also [wrote a great demo](https://anirudhdagar.ml/array-api-demo/GW_Demo_Array_API.html) recreating gravitational waves tutorial and showcasing what it would take to support PyTorch tensors for a subset of SciPy API using `__array_namespace__` .
+
+## PyData dispatching system
+
+In the future, we'd like to pass alternative arrays to SciPy and scikits functions and get the results without transforming internally to NumPy arrays. Instead, it would automatically handle dispatching to the appropriate implementation, for example, `cupyx.scipy` for GPU-accelerated SciPy functions, scikit-learn-intelex for Intel optimized version of scikit-learn, or cuCIM for GPU-accelerated scikit-image.
+Usually in statically typed languages function overloading is used to implement a function with the same name but arguments of different types. How can we implement something similar to functions overloading but in Python? Well, we could implement one central function checking the types of the arguments using `isinstance()`, or `issubclass()` and calling appropriate implementations based on these checks. This approach doesn't scale well with the number of inputs of different types and is not extendable by third-party implementers without modifying the source code.
```

Comment:
I like that we mention this approach here since this is the first thing that comes to mind even though it is not very scalable and efficient. Would it be a good idea to also mention a real world example like the [einops](https://github.com/arogozhnikov/einops) example here since they particularly use this approach to manage backends?

Reply:
@IvanYashchuk @rgommers what do you think? I guess this comment got buried under other small review comments.

Reply:
Maybe useful indeed, but very briefly. For example "(see `einops.backends` as an example)" (with that being a link).

Reply:
Added linking to https://github.com/arogozhnikov/einops/blob/1225f50eebd297f82a3fb97fca87319b522cd99c/einops/_backends.py#L22-L29. See if you agree @IvanYashchuk

Reply:
Thank you, Anirudh, for the suggestion, and thank you, Ralf, for adding the link.